torch/testing/_internal/common_utils.py

```diff
@@ -803,204 +810,412 @@ def is_integral(self, dtype):
+            count_non_identical = torch.sum(identity_mask, dtype=torch.long)
+            diff = torch.abs(a_flat - b_flat)
+            greatest_diff_index = torch.argmax(diff)
+            debug_msg = ("Found {0} different element(s) (out of {1}), with the greatest "
+                         "difference of {2} ({3} vs. {4}) occuring at index "
+                         "{5}.".format(count_non_identical.item(),
+                                       a.numel(),
+                                       diff[greatest_diff_index],
+                                       a_flat[greatest_diff_index],
+                                       b_flat[greatest_diff_index],
+                                       self._unravel_index(greatest_diff_index, a.shape)))
+            return (False, debug_msg)
+
+        # Compares complex tensors' real and imaginary parts separately.
+        # Note: this is intentionally divergent from complex torch.isclose
+        # which implements a more mathematically natural notion of "closeness"
+        # for complex numbers.
+        if a.is_complex():
+            float_dtype = torch.float32 if a.dtype == torch.complex64 else torch.float64
+            a_real = a.copy_real().to(float_dtype)
+            b_real = b.copy_real().to(float_dtype)
+            real_result, debug_msg = self.compareTensors(a_real, b_real, rtol=rtol, atol=atol,
+                                                         equal_nan=equal_nan, exact_dtype=exact_dtype,
+                                                         exact_device=exact_device)
+
+            if not real_result:
+                debug_msg = "Real parts failed to compare as equal! " + debug_msg
+                return (real_result, debug_msg)
+
+            a_imag = a.copy_imag().to(float_dtype)
+            b_imag = b.copy_imag().to(float_dtype)
+            imag_result, debug_msg = self.compareTensors(a_imag, b_imag, rtol=rtol, atol=atol,
+                                                         equal_nan=equal_nan, exact_dtype=exact_dtype,
+                                                         exact_device=exact_device)
+
+            if not imag_result:
+                debug_msg = "Imaginary parts failed to compare as equal! " + debug_msg
+                return (imag_result, debug_msg)
+
+            return (True, None)
+
+        # All other types use torch.allclose directly
+        if torch.allclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan):
+            return (True, None)
+
+        # Gathers debug info for failed float tensor comparison
+        a_flat = a.to(torch.float64).flatten()
+        b_flat = b.to(torch.float64).flatten()
+        diff = torch.abs(a_flat - b_flat)
+        outside_range = diff > (atol + rtol * torch.abs(b_flat))
+        count_outside_range = torch.sum(outside_range, dtype=torch.long)
+        greatest_diff_index = torch.argmax(diff)
+        debug_msg = ("With rtol={0} and atol={1}, found {2} element(s) (out of {3}) whose "
```

Comment:
it still seems to me like it would be preferable to re-use the implementation we have in torch.assert_allclose, instead of creating/testing/maintaining a new one with similar debug info. There's a lot of code here and it seems like it does stuff we already have elsewhere (and which could probably also be improved)

Reply:
I did crib from that implementation when implementing this. They follow the same pattern but I updated the message for context. The one thing this doesn't do is mask out NaNs / infs. I could add that but I thought it was a little misleading? We could update the error messages to be something like, "There were infinite differences but the greatest finite difference was..." But that seems like a lot. 

I agree two implementations isn't ideal, though, and assert_allclose's behavior is divergent, too. I could update assert_allclose to call compareTensors?

Reply:
We can always mask nans and then call allclose or something like that (I think that's what master does?), I'm fine with merging the implementations too, I just want to make sure we're leaving this in a maintainable state.

> "There were infinite differences but the greatest finite difference was..."

if allow_infs is false, detecting any inf value just means assertion fails. if we allow_infs, they have to match (have the same sign). I don't think the message has to contain information about both inf and particular values if the tensor contains differences in both.

Reply:
I updated assert_allclose to rely on a common function and improved the testing, including cases where the tensors being compared have extremal values.

allow_infs is always ignored on current master except when comparing scalars, and this PR removes the kwarg entirely for clarity. 

assert_allclose doesn't actually mask NaNs in its current implementation. It does mask close values, and differences in close values might result in NaNs (e.g. inf - inf or NaN - NaN), but it doesn't actually have special handling for cases where a NaN is being compared to a non-NaN value. I added special NaN case to the debug message.

Old message example:

```
a = torch.tensor((5, float('nan')))
b = torch.tensor((5, 7), dtype=torch.float32)
torch.testing.assert_allclose(a, b)
: AssertionError: Not within tolerance rtol=0.0001 atol=1e-05 at input[1] (nan vs. 7.0) and 0 other locations (50.00%)
```

Test for new message behavior:

```
a = torch.tensor((20, -6), device=device, dtype=torch.float32)
b = torch.tensor((-1, float('nan')), device=device, dtype=torch.float32)
result, debug_msg = self._compareTensors(a, b)
expected_msg = ("With rtol=1.3e-06 and atol={0}, found 2 element(s) (out of 2) "
                "whose difference(s) exceeded the margin of error (including 1 nan comparisons). "
                "The greatest difference was nan (-6.0 vs. nan), "
                "which occurred at index 1.").format(atol)
self.assertEqual(debug_msg, expected_msg)
```

In both cases the comparison to nan is cited as the "greatest" difference, although the new message highlights how many nan comparisons there are.